### ISSUE TITLE
[FEATURE] Add option to disable automatic image padding on dataloader

### ISSUE THREAD

**Issue by alphinside** [nodefluxio/vortex#48]
**Is your feature request related to a problem? Please describe.**
On issue #43 , DETR training involved their own mechanism to pad the image and building the mask. Somehow having this option means we can follow its mechanism

**Describe the solution you'd like**
Still looking for some ways developer can provide information to 'create_dataset' function

**Comment by alifahrri** [comment]
do you mean image padding or batch padding? doesn't torch dataloader padding behaviour depends on collate_fn? @alphinside

**Comment by alphinside** [comment]
image padding, in current implementation default behavior is automatic padding to square, so in collate_fn you didn't need to resize or padding anymore in order to make torch stack of image data

**Comment by alifahrri** [comment]
doesn't image padding performed on augments/standard_augments in wrapper, not dataloader?

**Comment by triwahyuu** [comment]
it is in dataset wrapper actually, probably you mean in create_dataloader or create_dataset?

**Comment by alphinside** [comment]
yes, correct, however note that we also have DALILoader in which the augmentation is not done on the dataset object

**Comment by alphinside** [comment]
note that in PytorchDataLoader, augmentations is in dataset object, in DALIDataLoader, augmentations is in pipeline

**Comment by triwahyuu** [comment]
so it's in create_dataloader, right?

**Comment by alphinside** [comment]
> note that in PytorchDataLoader, augmentations is in dataset object, in DALIDataLoader, augmentations is in pipeline

but after looking it more closely, yes we can put the arguments in the dataset object

**Comment by alphinside** [comment]
> so it's in create_dataloader, right?

subject to discuss cc @alifahrri

**Comment by alifahrri** [comment]
well it think it should be in create_dataset, putting image padding in create_dataloader doesn't seem right

**Comment by alphinside** [comment]
anyway what have been discussed until this point is implementation details which is quite straightforward, the discussion subject is more about the origin of this disable_auto_pad value. In my proposal I prefer it put on the preprocess_args when the developer develop the model, so the developer fully aware that he is trying to disable the autopad and must develop his own mechanism on the collate_fn cc @alifahrri  @triwahyuu

**Comment by alphinside** [comment]
> well it think it should be in create_dataset, putting image padding in create_dataloader doesn't seem right

but how we forward the information to create dataset? considering it is called inside the create_dataloader

**Comment by triwahyuu** [comment]
I'm fine with your proposal

**Comment by alifahrri** [comment]
> > well it think it should be in create_dataset, putting image padding in create_dataloader doesn't seem right
> 
> but how we forward the information to create dataset? considering it is called inside the create_dataloader

wait, turns out we have create_dataloader and create_loader

but yeah preprocess_args is visible to dataset

**Comment by alphinside** [comment]
After I look closely, `preprocess_args` is not returned from the model_components. Its Normalize() object and it's difficult to find place to originate this value. I prefer not to expose this option to the user, only to developer. So i change my suggestion to embed this value as an atrributes in the `collate_fn` object

**Comment by alphinside** [comment]
Actually `preprocess_args` is not modified by the `model_components`. So my previous proposal is invalid. So yeah still looking an idea on how to forward information from developer to the `create_dataset` function

**Comment by alifahrri** [comment]
it is not returned, but it is mutable btw

**Comment by alphinside** [comment]
> it is not returned, but it is mutable btw

yeah you're correct

**Comment by triwahyuu** [comment]
so it is still possible right?

**Comment by alphinside** [comment]
> so it is still possible right?

unfortunately no, currently there's no way to forward information from developer's side and maintain a good sensible arguments. To achieve this, we need to move out the `create_dataset` and `create_collater` out of `create_dataloader` and from that point we can modify that.

**Comment by alphinside** [comment]
So i think this is my `workaround` proposal :

```
    dataset = create_dataset(dataset_config=dataset_config, 
                             stage=stage, 
                             preprocess_config=preprocess_config,
                             wrapper_format=wrapper_format[dataloader_module])
    if isinstance(collate_fn,str):
        collater_args = {}
        try:
            collater_args = dataloader_config.collater.args
        except:
            collater_args = {}
        collater_args['dataformat'] = dataset.data_format
        collate_fn = create_collater(collate_fn, **collater_args)

        # Re-initialize dataset (Temp workaround)
        if hasattr(collate_fn,'disable_image_auto_padding'):
            dataset = create_dataset(dataset_config=dataset_config, 
                                     stage=stage, 
                                     preprocess_config=preprocess_config,
                                     wrapper_format=wrapper_format[dataloader_module],
                                     disable_image_auto_padding=collate_fn.disable_image_auto_padding)
```
